### ISSUE TITLE
\r and ANSI colors produce wrong output, sometimes

### ISSUE THREAD

**Issue by gaborcsardi** [rstudio/rstudio#7278]
### System details

    RStudio Edition : Desktop
    RStudio Version : 1.4.530
    OS Version      : macOS Mojave
    R Version       : 4.0.0

It seems to be present in 1.3.959 (Desktop) and 1.2.5042-1 (via RStudio Cloud) as well.

### Steps to reproduce the problem

In the console:

```r
str <- c(
  "B", "\033[31m", "x", "\033[39m", "A",
  "\r   ",
  "\rMessage\n"
)
cat(str, sep = "")
```

The (final) output is three spaces. The expected final output is this:

```
Message
```

The terminal works corrently.

### Describe the problem in detail

1. In `str`, we create a string of three characters, the middle one is colored red with an ANSI control sequence. 
2. Then the `"\r   "` string should move the cursor back to the beginning and overwrite the `BxA` string.
3. Then we move the cursor back again, and print a message.

### Describe the behavior you expected

The output should be:

```
Message
```

- [x] I have read the guide to submitting good bug reports at https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports . 
- [x] I have installed the latest version of RStudio and confirmed that the issue still persists.
- [x] If I am reporting a RStudio crash, I have included a diagnostics report. https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by gtritchie** [comment]
These are relatively safe fixes (this area has lots of unit-test coverage).

**Comment by gaborcsardi** [comment]
> These are relatively safe fixes (this area has lots of unit-test coverage).

Sounds great to me. :)

**Comment by gaborcsardi** [comment]
Thanks!

**Comment by R-Hannibal** [comment]
Verified Fixed on: RStudio Version - 1.4.747, R Version - 4.0.2, Windows 10 x64 - Desktop

Was able to reproduce the issue on stable 1.3 (1.3.1073), fix confirmed successful on the latest Daily (*the reported string is now outputting Message when run in the console or from an R Script).
